eval-0trace/parse-each.go

```diff
@@ -0,0 +1,209 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"flag"
+	"github.com/go-ping/ping"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const zeroTraceBeginString = "TRACE RESULTS"
+
+var (
+	InfoLogger *log.Logger
+	EvalLogger *log.Logger
+)
+
+type Hops struct {
+	Hop int
+	IP  string
+}
+
+type Result struct {
+	Timestamp  string
+	ClientIP   string
+	Traceroute []Hops
+	Lasthop    string
+	Lastping   RtItem
+	ClientPing RtItem
+}
+
+type AvgRTTcompare struct {
+	Timestamp  string
+	ClientIP   string
+	LastHop    float64
+	ClientHop  float64
+	Difference float64
+}
+
+const ICMPCount = 5
+const ICMPTimeout = time.Second * 10
+
+type RtItem struct {
+	IP        string
+	PktSent   int
+	PktRecv   int
+	PktLoss   float64
+	MinRtt    float64
+	AvgRtt    float64
+	MaxRtt    float64
+	StdDevRtt float64
+}
+
+// IcmpPing conducts the ICMP ping measurement, code from proxy-detection-server
+func IcmpPing(ip string) (RtItem, error) {
+	pinger, err := ping.NewPinger(ip)
+	if err != nil {
+		log.Println(err)
+		return RtItem{}, err
+	}
+	pinger.Count = ICMPCount
+	pinger.Timeout = ICMPTimeout
+	err = pinger.Run() // Blocks until finished.
+	if err != nil {
+		log.Println(err)
+		return RtItem{}, err
+	}
+	stat := pinger.Statistics()
+	icmp := RtItem{ip, stat.PacketsSent, stat.PacketsRecv, stat.PacketLoss, fmtTimeMs(stat.MinRtt), fmtTimeMs(stat.AvgRtt), fmtTimeMs(stat.MaxRtt), fmtTimeMs(stat.StdDevRtt)}
+	return icmp, nil
+
+}
+
+// calcAvgDifference calculates the difference between the AvgRtt of the two given RtItem objects
+func calcAvgDifference(last RtItem, client RtItem) (float64, bool) {
+	if last.AvgRtt == 0 || client.AvgRtt == 0 {
+		return 0, false
+	}
+	return client.AvgRtt - last.AvgRtt, true
+}
+
+func fmtTimeMs(value time.Duration) float64 {
+	return (float64(value) / float64(time.Millisecond))
+}
+
+// parseTracerouteFile takes the path of a 0trace output file as input
+// and parses and returns the traceroute hops, clientIP from the path, and the last hop obtained by 0trace
+func parseTracerouteFile(tracefile string) ([]Hops, string, string) {
```

Comment:
Personal opinion: I find it cleaner to return a struct instead of multiple fields at once. In this case, I would have done something like the following:
```go
type ZeroTraceRun struct {
  Hops []Hop
  ClientAddr string
}

func parseTracerouteFile(tracefile string) *ZeroTraceRun {
  ...
}
```
(The last hop is already part of `[]Hop`, so I didn't add an extra field for that.)

Anyway, no need to change this; just figured I'd let you know!

Reply:
Right, that is neater. Will keep in mind for the future.